File: 2023/Leetcode/problem141.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    if (!head || !head.next || !head.next.next) {
        return false;
    }
    
    let fast = head;
    let slow = head;

    do {
        fast = fast.next.next;
        slow = slow.next;
    } while (fast !== slow && fast.next && fast.next.next);

    return fast === slow;
};

File: 2023/Leetcode/problem141.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next or not head.next.next:
            return False

        slow = head.next
        fast = head.next.next
        while fast.next and fast.next.next and slow != fast:
            slow = slow.next
            fast = fast.next.next

        return slow == fast